packages/yarnpkg-pnpify/sources/NodeModulesFS.ts

```diff
@@ -34,15 +56,50 @@ type PortableNodeModulesFSOptions = {
   baseFs?: FakeFS<PortablePath>
 };
 
+type WatchedDirMap = Map<PortablePath, WatchedDir>;
+
+interface WatchedDir {
+  lastWatcherId: number;
+  watchers: { [watcherId: number]: Watcher & EventEmitter };
+  entries: Filename[];
+}
+
 class PortableNodeModulesFs extends FakeFS<PortablePath> {
   private readonly baseFs: FakeFS<PortablePath>;
-  private readonly pathResolver: NodePathResolver;
+  private readonly watchedDirs: WatchedDirMap = new Map();
+  private pathResolver: NodePathResolver;
 
   constructor(pnp: PnpApi, {baseFs = new NodeFS()}: PortableNodeModulesFSOptions = {}) {
     super(ppath);
 
     this.baseFs = baseFs;
     this.pathResolver = new NodePathResolver(pnp);
+    const pnpFilePath = npath.join(pnp.getPackageInformation(pnp.topLevel)!.packageLocation, '.pnp.js');
+    this.watchPnpFile(pnpFilePath);
+  }
+
+  private watchPnpFile(pnpFilePath: NativePath) {
+    this.baseFs.watch(NodeFS.toPortablePath(pnpFilePath), () => {
+      delete require.cache[pnpFilePath];
+      const pnp = require(pnpFilePath);
+      // Sometimes pnp object is empty, which is weird, but eventually it comes correctly filled
```

Comment:
The problem has gone after switching to directory watching, instead of `.pnp.js` file watching. Direct `.pnp.js` file watching was not reliable for me anyways, so I had to switch to containing directory watching to detect changes to `.pnp.js`